File: src/euler7.rs
mod numthry;
use numthry::Primes;

/// What is the 10 001st prime number?
fn main() {
	// note that the first item is .nth(0)
	println!("{}", Primes::new().nth(10_001 - 1).unwrap());
}

/// By listing the first six prime numbers: 2, 3, 5, 7, 11, and 13, we can see that the 6th prime is 13.
#[test]
fn test() {
	let expected = vec![2, 3, 5, 7, 11, 13];
	let result: Vec<u64> = Primes::new().take(6).collect();
	assert_eq!(result, expected);
}

File: c/src/euler7.c
//prepare: -lm
#include <stdio.h>
#include <stdbool.h>
#include <assert.h>
#include <math.h>
#include "number_theory.c"

/*
By listing the first six prime numbers: 2, 3, 5, 7, 11, and 13, we can see that the 6th prime is 13.
What is the 10 001st prime number?
*/

unsigned nth_prime(unsigned n){
	if (n == 1){
		return 2;
	}
	unsigned candidate = 3;
	unsigned primes_found = 2;
	while (primes_found < n){
		candidate += 2;
		if (is_prime(candidate)){
			primes_found += 1;
		}
	}
	return candidate;
}

int main(){
	assert(nth_prime(6) == 13);
	printf("%u\n", nth_prime(10001));
	return 0;
}